src/lib/HST/Effect/Report.hs

```diff
@@ -54,13 +58,161 @@ import           HST.Effect.Cancel ( Cancel, cancel, runCancel )
 data Severity = Internal | Error | Warning | Info | Debug
  deriving ( Show, Eq )
 
--- | A messages that can be 'report'ed.
-data Message = Message { msgSeverity :: Severity, msgText :: String }
+-- | A message that can be 'report'ed.
+data Message = Message { msgSeverity :: Severity
+                       , msgSrcSpan  :: Maybe S.MsgSrcSpan
+                       , msgText     :: String
+                       }
  deriving ( Show, Eq )
 
+-- | Like using the 'Message' constructor directly, but takes a 'S.SrcSpan'
+--   instead of a @Maybe@ 'S.MsgSrcSpan' and does the conversion.
+message :: Severity -> S.SrcSpan a -> String -> Message
+message severity srcSpan = Message severity (S.toMsgSrcSpan srcSpan)
+
 -- TODO Add @Pretty@ instance for messages.
+-- | Pretty-prints a message without a code excerpt.
+--
+--   Only used for testing purposes where the 'HST.Effect.InputFile' effect is
+--   not always handled.
 showPrettyMessage :: Message -> String
-showPrettyMessage (Message severity msg) = show severity ++ ": " ++ msg
+showPrettyMessage (Message severity Nothing msg)
+  = show severity ++ ": " ++ msg
+showPrettyMessage (Message severity (Just src) msg)
+  = show severity ++ ": " ++ msg ++ "\n  In " ++ prettyMsgSrcSpan src ++ "."
+
+-- | Pretty-prints a message with a code excerpt.
+showPrettyMessageWithExcerpt :: Member InputFile r => Message -> Sem r String
+showPrettyMessageWithExcerpt (Message severity srcSpan msg) = do
+  excerpt <- displayCodeExcerpt srcSpan
+  return
+    $ show severity
+    ++ ": "
+    ++ msg
+    ++ if null excerpt then "" else '\n' : excerpt
+
+-- | Builds a string displaying the file path and the line and column of the
+--   start and end of the given message source span.
+prettyMsgSrcSpan :: S.MsgSrcSpan -> String
+prettyMsgSrcSpan src
+  = S.msgSrcSpanFilePath src ++ ':' : prettyMsgSrcSpanNumbers src
+
+-- | Builds a string displaying the line and column of the given message source
+--   span start and end.
+prettyMsgSrcSpanNumbers :: S.MsgSrcSpan -> String
+prettyMsgSrcSpanNumbers src = show (S.msgSrcSpanStartLine src)
+  ++ ':'
+  : show (S.msgSrcSpanStartColumn src)
+  ++ '-'
+  : show (S.msgSrcSpanEndLine src) ++ ':' : show (S.msgSrcSpanEndColumn src)
+
+-- TODO Should the following function be in its own module? It can't be put
+-- back into 'HST.Frontend.Transformer.Messages' because of cyclic imports.
```

Comment:
I think this new module should include the functions in `HST.Frontend.Transformer.Messages` and be located in the `Util` folder. What do you think?

Reply:
The `Util` folder seems to be a good choice for the new module but I would recommend to keep the modules separate. Otherwise there is a cyclic dependency since the `HST.Frontend.Transformer.Messages` module uses the `Report` effect but `HST.Effect.Report` needs to import the new module.